administrator/components/com_jedchecker/libraries/rules/framework.php

```diff
@@ -149,6 +149,13 @@ protected function find($file)
 
 		$result = false;
 
+		if (preg_match('/<\?\s/', $content, $match, PREG_OFFSET_CAPTURE))
+		{
+			$lineno = substr_count($content, "\n", 0, $match[0][1]);
+			$this->report->addError($file, JText::_('COM_JEDCHECKER_ERROR_FRAMEWORK_SHORT_PHP_TAG'), $lineno + 1, $origContent[$lineno]);
```

Comment:
This should just be a warring, not an error as this is a recommendation.

Reply:
Note that `php.ini-production` file from PHP distro has `short_open_tag = Off`, and so any attempt to use `<?` on most production servers will result in incorrect behavior. So, I'd recommend to keep it as error-level message.

Reply:
We are not the code police, we only set errors messages where the component can not be listed on the JED because it violate a JED rule. So even if we know this is not good practice, its not and error level issue. We are freedom software, which must mean that we want to give the developers a heads-up on good practice, but we not going to force it.

Reply:
Then, why missed JEXEC guard is considered as error? Or direct work with database?

IMHO, JEDChecker should display errors in the following cases:
1. Violation of JED rules (e.g. incorrect name).
2. Issues that prevent extension from being installed (e.g. invalid manifest).
3. Issues that lead to incorrect work of extension (e.g. invalid language file).

Usage of `<?` comes to the latter category, because it means that the extension is not compatible with a default server settings.

If you disagree, let's make clear rules to distribute issues between error, warning, compatibility, and info levels, and reconsider levels of current issues.

Reply:
This is a new addition that you are proposing, so we are not talking about `JEXEC` we are talking about `<?` short hand that is commonly used in the HTML areas like this `<? $value ?>` that means the same as `<?php echo $value; ?>`. This is not seen as bad practice and is used by many developers in and outside of Joomla.

To be part of the conversation directly in relation to when what status a message should be, you are welcome to become an official member of the JED team, and raise you voice in our official meetings, and help set protocol. We need people like you to step-up and help us manage all these important responsibilities.

But from a normal developers perspective I don't agree this the use of `<?` instead of `<?php` in your code disqualify you from the JED listing. (since that is what an error should mean) This has never been the case, and pushing this change will require more motivation.

Saying that the default of PHP `php.ini-production` is set to `short_open_tag = Off` does not make that case that this is industry standard, as any shared host administrator like myself will openly admit that the defaults are never enough, and we all tweak out systems to best server our clients, and clearly the '<?' is no security risk, or performance penalty.

That means I could have said... no this pull request is un-needed, and just closed it. But instead I thought... giving developers a heads-up and setting a warning is okay as an initial step toward the future... where this short hand may be dropped (who knows). The moment that happens (not that I think it will) we move this message up to an error since then we are fully entitled to do so. At the moment seems premature to me.

Never forget we are on the same side... we want Joomla extensions developers to have as much freedom as we can possibly afford them. Yet at the same time, we want to protect those who use the extensions listed. By this we can't guarantee quality, or anything really.

We do not want to make it harder for developers, and if that has been your experience then please except my apology on behave of the JED. Remember JED team leaders are more like facilitators of the team, we protect, advice, and serve. We are often unable to bring change, or prevent change. So its like guiding running water... we try to get the water where it should go without losing any along the way.. we try.

Reply:
Note that shorthand echo (`<?=`) is allowed, AFAIK there is no way disable it in PHP, but `<?` (shorthand alternative to `<?php`) is disabled by default on prod servers.

And it's not correct to say "`<?` is no security risk", because mix of `<?php` and `<?` may lead to issues that is hard to found, e.g.:
- parsing errors resulting in paths disclosure in the case of `display_errors` turned on, or
- bypassing some auth checks (e.g. `<? if ($smth_allowed): ?> code visible for admins only <? endif; ?>`).

Reply:
A few links in support of error-level for short open PHP tags detection:

1.  This tag is not allowed by [Joomla's Coding Standard](https://developer.joomla.org/coding-standards/php-code.html):

    > Always use the full `<?php ... ?>` to delimit PHP code, not the `<? ... ?>` shorthand. This is the most portable way to include PHP code on differing operating systems and setups.

    Of course, extensions can use any coding standard (or none), but note the key point is **portability**, and extensions uploaded to JED are considered to be portable.

2.  Recently, in 2019, there was an [PHP-RFC to deprecate and remove short open PHP tags](https://wiki.php.net/rfc/deprecate_php_short_tags_v2) from PHP at all:

    > First, `short_open_tag` is an ini setting that control core language syntax. This means that their use is not possible in portable code, because the code author does not necessarily have the necessary control over the configuration of the deployment environment. Worse than that, code using short open tags deployed on a server using `short_open_tag=0` will leak application code, because short open tags are silently ignored.

    It got over 1/2 of supporting votes, but didn't pass required 2/3 votes. The conclusion was that in spite of the fact that short open PHP tags are not portable, it's better to keep them for backward compatibility, especially taking into account that they are disabled in all known coding standards (e.g. it is the first rule in the [PSR-1 Standard](https://www.php-fig.org/psr/psr-1/)).

So, all coding standards require to use `<?php` instead of `<?` for portability, and seems there is no exception.

Reply:
HI, I just reviewed an extension with this short code in it and it threw me off since red means no jed listing. It either should be a warning. Or we need to first edit our policy for the JED and only then make it an error. 
Like llewellyn said. Errors should mean no listing. And altho i agree it should be something that highly recommends to be fixed. It should not be an error if it means we're still going to publish the extension.

Reply:
@TLWebdesign The problem is that `short_open_tag` is disabled in most PHP setups (see default php.ini-development and php.ini-production files in PHP sources, default cli/php.ini and fpm/php.ini in debian/ubuntu repos and ppa:ondrej/php, etc.). IMHO, if an extension description states clearly that `short_open_tag` should be set `On` for this extension to work properly, it could be accepted, but I don't know context where it might be really necessary and preferred.